beacon_node/beacon_chain/src/data_availability_checker.rs

```diff
@@ -511,6 +512,114 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
             block_cache_size: self.availability_cache.block_cache_size(),
         }
     }
+
+    pub fn reconstruct_data_columns(
+        &self,
+        block_root: &Hash256,
+    ) -> Result<Option<AvailabilityAndReconstructedColumns<T::EthSpec>>, AvailabilityCheckError>
+    {
+        // Clone the pending components, so we don't hold the read lock during reconstruction
+        let Some(pending_components) = self
+            .availability_cache
+            .peek_pending_components(block_root, |pending_components_opt| {
+                pending_components_opt.cloned()
+            })
+        else {
+            // Block may have been imported as it does not exist in availability cache.
+            return Ok(None);
+        };
+
+        if !self.should_reconstruct(&pending_components) {
+            return Ok(None);
+        }
+
+        self.availability_cache
+            .set_reconstruction_started(block_root);
```

Comment:
There's a potential race condition here. We clone an old view of the pending components, check if `reconstruction_started`, then acquire the lock again in `set_reconstruction_started` and set `reconstruction_started` to true. We could have N parallel reconstruction attempts happening:

- thread 1: peek_pending_components + clone + drop lock
- thread 2: peek_pending_components + clone + drop lock
- thread 1:  `should_reconstruct` returns true
- thread 2:  `should_reconstruct` returns true

So either hold the lock until you set_reconstruction_started or restructure the check into being atomic.

Reply:
Good catch!

Reply:
I think this is fixed in https://github.com/sigp/lighthouse/pull/6403/commits/b0c337946de2a33c08eab8515d9612a27ca9c452.

I've also added error handling to reconstruction failure, so we can recover from potentially bad columns that somehow made it through to DA checker.